### PR TITLE
Remove run_on_reactor

### PR DESCRIPTION
**Pull request by hawkowl**

### CODE REVIEW

File: synapse/util/async.py [hunk 1]
```diff
@@ -227,7 +224,7 @@ def queue(self, key):
             # the context manager, but it needs to happen while we hold the
             # lock, and the context manager's exit code must be synchronous,
             # so actually this is the only sensible place.
-            yield run_on_reactor()
+            yield self._clock.sleep(0)
```
Comment (richvdh) [inline synapse/util/async.py]: can we make this do `sleep(0)` for now, and then replace it with `self._clock.sleep` in the PR which does the sleep replacement?

File: synapse/util/__init__.py [hunk 2]
```diff
@@ -31,16 +30,24 @@ def unwrapFirstError(failure):
...
-    mocked during unit-tests.
-
-    TODO(paul): Also move the sleep() functionality into it
     """
+    A Clock wraps a Twisted reactor and provides utilities on top of it.
+    """
+    _reactor = attr.ib(default=reactor)
+
+    @defer.inlineCallbacks
+    def sleep(self, seconds):
```
Comment (richvdh) [inline synapse/util/__init__.py]: can we push this to a separate PR?

File: synapse/util/async.py (hunk 1 above)
Comment (hawkowl) [reply synapse/util/async.py]: Hmm, yeah. Don't think I meant to cherry pick this out. Fixed.

**Comment by hawkowl** [comment]
@richvdh the clock stuff I think was because of ObservableDeferred, didn't realise I could just directly use the sleep that wasn't there... oops.

Now it's just the removal of `run_on_reactor`'s uses, the function itself, and ObservableDeferred directly using `sleep()` for now.

**Pull request merged**